### PR TITLE
QoS & Parallelism Level per label configuration

### PR DESCRIPTION
**Pull request by semion-akimtsev**
This update enables QoS and parallelism level configuration for a specific incoming message flow. If these settings are not specified the endpoint configuration is used.

### CODE REVIEW

File: Sources/Contour/Configurator/AppConfigConfigurator.cs [hunk 1]
```diff
@@ -199,6 +201,9 @@ public IBusConfigurator Configure(string endpointName, IBusConfigurator cfg)
                     cfg.RegisterValidator(v);
                 }
             }
+            #endregion
```
Comment (mzabolotko) [inline Sources/Contour/Configurator/AppConfigConfigurator.cs:204]: Remove #region and #endregion from sources next time.

File: Sources/Contour/Configurator/AppConfigConfigurator.cs [hunk 2]
```diff
@@ -230,36 +235,83 @@ public IBusConfigurator Configure(string endpointName, IBusConfigurator cfg)
...
+
+                //this should be the default value
+                var qos = configurator.GetQoS();
+
+                if (qos.HasValue)
+                {
+                    var size = qos.Value.PrefetchSize;
+                    var count = qos.Value.PrefetchCount;
+                    
+                    if (endpointConfig.Qos.PrefetchSize.HasValue)
```
Comment (mzabolotko) [inline Sources/Contour/Configurator/AppConfigConfigurator.cs]: Can QoS per label be set only if QoS per endpoint was set?

File: Sources/Contour/Transport/RabbitMQ/BusConsumerConfigurationEx.cs [hunk 3]
```diff
@@ -38,6 +40,17 @@ public static IReceiverConfigurator WithQoS(this IReceiverConfigurator builder,
         }
 
         /// <summary>
+        /// Returns current QoS settings
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <returns></returns>
+        public static Maybe<QoSParams> GetQoS(this IReceiverConfigurator builder)
+        {
+            var qos = ((RabbitReceiverOptions)((ReceiverConfiguration)builder).Options).GetQoS();
```
Comment (mzabolotko) [inline Sources/Contour/Transport/RabbitMQ/BusConsumerConfigurationEx.cs]: Maybe<T> pattern is used to escape needless exceptions. Check type safety and use Maybe<T> to return value.

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs [hunk 4]
```diff
@@ -1078,6 +1079,150 @@ public void should_route_any()
...
+
+                var result = configurator.Configure(endpointName, busConfiguration);
+                var busConfigurationResult = (BusConfiguration) result;
+                var receiverConfiguration = busConfigurationResult.ReceiverConfigurations.First();
+                var receiverOptions = (RabbitReceiverOptions) receiverConfiguration.Options;
+                
+                var qos = receiverOptions.GetQoS();
+                var value = qos.Value;
+
+                value.PrefetchCount.Should().Be(prefetchCount, "Incoming QoS prefetch count should be set");
```
Comment (mzabolotko) [inline Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs:1120]: The test checks two cases.

File: Sources/Contour/Configurator/AppConfigConfigurator.cs (hunk 1 above)
Comment (semion-akimtsev) [reply Sources/Contour/Configurator/AppConfigConfigurator.cs:204]: Do we have any coding guidelines on this?

Comment (alexzaitzev) [reply Sources/Contour/Configurator/AppConfigConfigurator.cs:204]: Yes. Do not use regions)

File: Sources/Contour/Configurator/AppConfigConfigurator.cs (hunk 2 above)
Comment (semion-akimtsev) [reply Sources/Contour/Configurator/AppConfigConfigurator.cs]: Yes and this should be changed. But it should only be available for RabbitMQ transport configuration.

File: Sources/Contour/Transport/RabbitMQ/BusConsumerConfigurationEx.cs (hunk 3 above)
Comment (semion-akimtsev) [reply Sources/Contour/Transport/RabbitMQ/BusConsumerConfigurationEx.cs]: Need some more explanation on this.

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 4 above)
Comment (semion-akimtsev) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs:1120]: What cases?

File: Sources/Contour/Transport/RabbitMQ/BusConsumerConfigurationEx.cs (hunk 3 above)
Comment (semion-akimtsev) [reply Sources/Contour/Transport/RabbitMQ/BusConsumerConfigurationEx.cs]: As discussed need to cast safely and check for null. Further refactoring needed in BusConsumerConfigurationEx.

File: Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs (hunk 4 above)
Comment (semion-akimtsev) [reply Tests/Contour.Configurator.Tests/BusConfigurationSpecs.cs:1120]: As discussed we need to split the test in two to check prefetch count and prefetch size separatelly.

**Pull request merged**